src/Runtime/jni/jniwrapper.c

```diff
@@ -595,37 +595,12 @@ jobject omtl_native_to_java(
 
     /* Create direct byte buffer Java object from native data buffer.
      *
-     * If jni_owning is true, we take ownership by setting owner flag
-     * to false. This means that when we call omTensorListDestroy
-     * the data buffer will not be freed since it has been given to
-     * the Java direct byte buffer and the Java GC will be responsible
-     * for freeing the data buffer. This way we avoid copying the data
-     * buffer.
-     *
-     * If jni_owning is false, it means the data buffer is not freeable
-     * due to one of the two following cases:
-     *
-     *   - user has malloc-ed the data buffer so the user is
-     *     responsible for freeing it
-     *   - the data buffer is static
-     *
-     * Either way, since the data buffer will be given to Java and is
-     * subject to GC, we must make a copy of the data buffer.
+     * When the Java OMTensor object is created, it will create a copy of the
+     * tensor data to ensure Java ownership of the underlying tensor data.
+     * Creation of the DirectByteBuffer here is merely a passthrough mechanism
+     * to allow for the copy to occur.
      */
     void *jbytebuffer_data = jni_data;
-    if (jni_owning) {
-      LIB_CALL(omTensorSetOwning(jni_omts[i], (int64_t)0), 1, env,
-          japi->jecpt_cls, "");
-      LOG_PRINTF(LOG_DEBUG, "omt[%d]:%p data %p ownership taken", i,
-          jni_omts[i], jni_data);
-    } else {
-      LIB_VAR_CALL(jbytebuffer_data, malloc(jni_bufferSize),
-          jbytebuffer_data != NULL, env, japi->jecpt_cls, "jbytebuffer_data=%p",
-          jbytebuffer_data);
-      memcpy(jbytebuffer_data, jni_data, jni_bufferSize);
-      LOG_PRINTF(LOG_DEBUG, "omt[%d]:%p data %p copied into %p", i, jni_omts[i],
-          jni_data, jbytebuffer_data);
-    }
```

Comment:
At the end of the loop, do we need to remove local references to the various java objects created?
```
diff --git a/src/Runtime/jni/jniwrapper.c b/src/Runtime/jni/jniwrapper.c
index 5fc6ee99..c703c4a0 100644
--- a/src/Runtime/jni/jniwrapper.c
+++ b/src/Runtime/jni/jniwrapper.c
@@ -658,6 +658,11 @@ jobject omtl_native_to_java(
     /* Set the OMTensor object in the object array */
     JNI_CALL(env, (*env)->SetObjectArrayElement(env, jobj_omts, i, jobj_omt), 1,
         NULL, "");
+
+    JNI_CALL(env, (*env)->DeleteLocalRef(env, jomt_data), 1, NULL, "");
+    JNI_CALL(env, (*env)->DeleteLocalRef(env, jomt_shape), 1, NULL, "");
+    JNI_CALL(env, (*env)->DeleteLocalRef(env, jomt_strides), 1, NULL, "");
+    JNI_CALL(env, (*env)->DeleteLocalRef(env, jobj_omt), 1, NULL, "");
   }

   /* Create the OMTensorList java object */
```
Normally local references are automatically released when the function goes out of scope. But here we are in a loop reusing all the object pointers for each tensor so I think we need these explicit DeleteLocalRef?

Reply:
While the C object pointers are being reused the refs are still tied to the scope of the function. Once the function returns back to Java, _any_ object that was created will have its local reference popped unless it was returned. Since these are transient objects I could add deletelocalref but given the size of the java objects they represent it only would matter if dealing with thousands of these within a single JNI call.

Reply:
OK I just want to make sure there is no leakage of the local reference. The loop could theoretically iterate over thousands of tensors (i.e., output tensor list has thousands of tensors). Would that be a problem without the DeleteLocalRef?

Reply:
It shouldn't be a problem. Remember, you're talking about saving maybe 1K per tensor, if even that much